src/order/well_founded_set.lean

```diff
@@ -499,6 +505,130 @@ end
+  s.partially_well_ordered_on r ↔
+    ∀ f, ¬ is_bad_seq r s f :=
+begin
+  rw [set.partially_well_ordered_on],
+  apply forall_congr (λ f, _),
+  simp [is_bad_seq]
+end
+
+/-- This indicates that every bad sequence `g` that agrees with `f` on the first `n`
+  terms has `rk (f n) ≤ rk (g n)`. -/
+def is_min_bad_seq (r : α → α → Prop) (rk : α → ℕ) (s : set α) (n : ℕ) (f : ℕ → α) : Prop :=
+  ∀ g : ℕ → α, (∀ (m : ℕ), m < n → f m = g m) → rk (g n) < rk (f n) → ¬ is_bad_seq r s g
+
+/-- Given a bad sequence `f`, this constructs a bad sequence that agrees with `f` on the first `n`
+  terms and is minimal at `n`.
+-/
+noncomputable def min_bad_seq_of_bad_seq (r : α → α → Prop) (rk : α → ℕ) (s : set α)
+  (n : ℕ) (f : ℕ → α) (hf : is_bad_seq r s f) :
+  { g : ℕ → α // (∀ (m : ℕ), m < n → f m = g m) ∧ is_bad_seq r s g ∧ is_min_bad_seq r rk s n g } :=
+begin
+  classical,
+  have h : ∃ (k : ℕ) (g : ℕ → α), (∀ m, m < n → f m = g m) ∧ is_bad_seq r s g
+        ∧ rk (g n) = k :=
+  ⟨_, f, λ _ _, rfl, hf, rfl⟩,
+  obtain ⟨h1, h2, h3⟩ := classical.some_spec (nat.find_spec h),
+  refine ⟨classical.some (nat.find_spec h), h1, by convert h2, λ g hg1 hg2 con, _⟩,
+  refine nat.find_min h _ ⟨g, λ m mn, (h1 m mn).trans (hg1 m mn), by convert con, rfl⟩,
+  rwa ← h3,
+end
+
+lemma exists_min_bad_of_exists_bad (r : α → α → Prop) (rk : α → ℕ) (s : set α) :
+  (∃ f, is_bad_seq r s f) → ∃ f, is_bad_seq r s f ∧ ∀ n, is_min_bad_seq r rk s n f :=
+begin
+  rintro ⟨f0, (hf0 : is_bad_seq r s f0)⟩,
+  let fs : Π (n : ℕ), { f :  ℕ → α // is_bad_seq r s f ∧ is_min_bad_seq r rk s n f },
+  { refine nat.rec _ _,
+    { exact ⟨(min_bad_seq_of_bad_seq r rk s 0 f0 hf0).1,
+        (min_bad_seq_of_bad_seq r rk s 0 f0 hf0).2.2⟩, },
+    { exact λ n fn, ⟨(min_bad_seq_of_bad_seq r rk s (n + 1) fn.1 fn.2.1).1,
+        (min_bad_seq_of_bad_seq r rk s (n + 1) fn.1 fn.2.1).2.2⟩ } },
+  have h : ∀ m n, m ≤ n → (fs m).1 m = (fs n).1 m,
+  { intros m n mn,
+    obtain ⟨k, rfl⟩ := exists_add_of_le mn,
+    clear mn,
+    induction k with k ih,
+    { refl },
+    rw [ih, ((min_bad_seq_of_bad_seq r rk s (m + k).succ (fs (m + k)).1 (fs (m + k)).2.1).2.1 m
+        (nat.lt_succ_iff.2 (nat.add_le_add_left k.zero_le m)))],
+    refl },
+  refine ⟨λ n, (fs n).1 n, ⟨set.range_subset_iff.2 (λ n, ((fs n).2).1.1 (mem_range_self n)),
+    λ m n mn, _⟩, λ n g hg1 hg2, _⟩,
+  { dsimp,
+    rw [← subtype.val_eq_coe, h m n (le_of_lt mn)],
+    convert (fs n).2.1.2 m n mn },
+  { convert (fs n).2.2 g (λ m mn, eq.trans _ (hg1 m mn)) (lt_of_lt_of_le hg2 (le_refl _)),
+    rw ← h m n (le_of_lt mn) },
+end
+
+lemma iff_not_exists_is_min_bad_seq {r : α → α → Prop} (rk : α → ℕ) {s : set α} :
+  s.partially_well_ordered_on r ↔ ¬ ∃ f, is_bad_seq r s f ∧ ∀ n, is_min_bad_seq r rk s n f :=
+begin
+  rw [iff_forall_not_is_bad_seq, ← not_exists, not_congr],
+  split,
+  { apply exists_min_bad_of_exists_bad },
+  rintro ⟨f, hf1, hf2⟩,
+  exact ⟨f, hf1⟩,
+end
+
+/-- Higman's Lemma -/
```

Comment:
I think this deserves a longer docstring that is readable by non-Lean-experts :stuck_out_tongue_winking_eye:

Reply:
Fair enough! Hopefully this new docstring is clear enough.